planning/MartenV4.md

```diff
@@ -0,0 +1,114 @@
+* Loading entities and tracking entities in the *identity map* or version tracking
+
+What could be generated in the future:
+
+* Document metadata properties -- but sad trombone, that might have to stay with Expressions if the setters are internal/private :/
+* Much more of the `ISelector` implementations, especially since there's going to be more variability when we do the document metadata
+* Finer-grained manipulation of the `IIdentityMap` 
+
+
+## Unit Testing Approach
+
+If we introduce the runtime code generation back into Marten, that's unfortunately a non-trivial "cold start" testing issue. To soften that, I suggest we get a lot more aggressive with reusable [xUnit.Net class fixtures](https://xunit.net/docs/shared-context) between tests to reuse generated code between tests, cut way down on the sheer number of database calls by not having to frequently check the schema configuration, and other `DocumentStore` overhead. 
+
+A couple other points about this:
+
+* We need to create more unique document types so we're not having to use different configurations for the same document type. This would enable more reuse inside the testing runtime
+* Be aggressive with separate schemas for different configurations
+* We could possibly turn on xUnit.net parallel test running to speed up the test cycles
+
+
+## Document Metadata
+
+* Use the configuration and tests from [pull request for Document Metadata](https://github.com/JasperFx/marten/pull/1364), but use the Lamar-backed dynamic code generation from the previous section to pull this off.
+* I think we could expand the document metadata to allow for user defined properties like "user id" or "transaction id" much the same way we'll do for the EventStore metadata. We'd need to think about how we extend the document tables and how metadata is attached to a document session
+
+
+## Project / Assembly Restructuring
+
+* Pull everything to do with Schema object generation and difference detection to a separate library (`IFeatureSchema`, `ISchemaObject`, etc.). Mostly to clean out the main library, but also because this code could easily be reused outside of Marten. Separating it out might make it easier to test and extend that functionality, which is something that occasionally gets requested. There's also the possibility of further breaking that into abstractions and implementations for the long run of getting us ready for Sql Server or other database engine support
+
+* *Possibly* pull the ADO.Net helper code like `CommandBuilder` and the extension methods into a small helper library somewhere else (I'm nominating the [Baseline](https://jasperfx.github.io/baseline) repository). This code is copied around to other projects as it is, and it's another way of getting stuff out of the main library and the test suite.
+
+* More in the event store section, but let's get the event store functionality into its own library (or possibly a couple other libraries for async projection support).
+
```

Comment:
We should also do a review of our code base and make everything private or internal that we feel shouldn't be the part of Marten public API. Currently, a lot of classes seem to have public modifier without much reason and it's not helping in maintenance.

I'd like to also make the Marten more modular - so eg. detach as much as possible the LINQ, ES, Document Storage, SQL generation into separate as much loosely coupled code parts.

Reply:
Agreed. I mostly want to slim down the Marten codebase and especially the Marten test suites

Reply:
In regards to going for the access modifiers to reduce the public surface, would it also make sense to address other code hygiene stuff? E.g. we have the namespace `Marten.Exceptions` with two exceptions but then there are also some other exceptions scattered around in different namespaces. And going with exceptions, should we have the standard exceptions ctors (https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1032?view=vs-2019)?

I've used NDepend with the Marten codebase from time to time... Could take a look at the most "offending" sites with the default rule set.

Reply:
I don't tend to give much credence to NDepend's view of "code quality" outside of the standard Cyclomatic Complexity kind of metrics, but we can look at the code organization. I'm dead set against doing any significant amount of work just to make the NDepend layering numbers look better. I wouldn't make the exception constructors a high priority, but it doesn't hurt either way IMO.

Reply:
@jokokko I have to admit that I'm not enthusiastic about these kinds of static code analysis rules, and I'm specifically dubious of NDepend's ideas of "code quality". I don't have a strong opinion about the exception constructors now that binary serialization and remoting is dead. I think gathering up the exceptions in a common namespace is okay, especially if they're meant for users.